tla-types/src/main/scala/at/forsyte/apalache/tla/typecheck/etc/TypeUnifier.scala

```diff
@@ -212,12 +217,107 @@ class TypeUnifier {
+
+      case (RecRowT1(RowT1(lfields, lv)), RecRowT1(RowT1(rfields, rv))) =>
+        unifyRows(lfields, rfields, lv, rv).map(t => RecRowT1(t))
+
+      case (VariantT1(RowT1(lfields, lv)), VariantT1(RowT1(rfields, rv))) =>
+        unifyRows(lfields, rfields, lv, rv).map(t => VariantT1(t))
+
       // everything else does not unify
       case _ =>
         None // no unifier
     }
   }
 
+  // unify two rows
+  @tailrec
+  private def unifyRows(
+      lfields: SortedMap[String, TlaType1],
+      rfields: SortedMap[String, TlaType1],
+      lvar: Option[VarT1],
+      rvar: Option[VarT1]): Option[RowT1] = {
+    // assuming that a type is either a row, or a variable, make it a row type
+    def asRow: Option[TlaType1] => Option[RowT1] = {
+      case Some(r @ RowT1(_, _)) => Some(r)
+      case Some(v @ VarT1(_))    => Some(RowT1(v))
+      case Some(tp)              => throw new IllegalStateException("Expected RowT1(_, _) or VarT1(_), found: " + tp)
+      case None                  => None
+    }
+
+    // consider four cases
+    if (lfields.isEmpty) {
+      // the base case
+      (lvar, rvar) match {
+        case (None, None) =>
+          if (rfields.nonEmpty) None else Some(RowT1())
+
+        case (Some(lv), Some(rv)) =>
+          if (rfields.isEmpty) {
+            asRow(compute(lv, rv))
+          } else {
+            asRow(unifyVarWithNonVarTerm(lv.no, RowT1(rfields, rvar)))
+          }
+
+        case (Some(lv), None) =>
+          asRow(unifyVarWithNonVarTerm(lv.no, RowT1(rfields, None)))
+
+        case (None, Some(rv)) =>
+          if (rfields.isEmpty) {
+            // the only way to match is to make the right variable equal to the empty row
+            asRow(unifyVarWithNonVarTerm(rv.no, RowT1()))
+          } else {
+            // the left row is empty, whereas the right row is non-empty
+            None
+          }
+      }
+    } else if (rfields.isEmpty) {
+      // the symmetric case above
+      unifyRows(rfields, lfields, rvar, lvar)
+    } else {
+      val sharedFieldNames = lfields.keySet.intersect(rfields.keySet)
+      if (sharedFieldNames.isEmpty) {
+        // The easy case: no shared fields.
+        // The left row is   (| lfields | lvar |).
+        // The right row is  (| rfields | rvar |).
+        // Introduce a fresh type variable to contain the common tail.
+        val tailVar = freshVar()
+        // Unify lvar with   (| rfields | tailVar |).
+        // Unify rvar with   (| lfields | tailVar |).
+        // If both unifiers exist, the result is (| lfields | rfields | tailVar |).
+        if (
+            compute(lvar.getOrElse(RowT1()), RowT1(rfields, Some(tailVar))).isEmpty
+            || compute(rvar.getOrElse(RowT1()), RowT1(lfields, Some(tailVar))).isEmpty
+        ) {
+          None
+        } else {
+          // apply the computed substitution to obtain the whole row
+          asRow(Some(Substitution(solution).sub(RowT1(lfields, lvar))._1))
```

Comment:
Well, if you don't need the substitution, it does unify things. But since we need the resulting substitution, `compute` mutates the substitution pretty much everywhere.